### PR TITLE
Add 'mimic' command

### PR DESCRIPTION
**Pull request by ampleyfly**
Mimic a user based on logs.
The command is implemented in the log service, which is arguably very ugly. Alternatives could be to send more data over dbus, or to have multiple modules access the database directly. I'm not sure these are better.

### CODE REVIEW

File: pladder/log.py [hunk 1]
```diff
@@ -140,6 +153,48 @@ def add_line(self, timestamp, channel, nick, text):
...
+                if prev != '':
+                    stats[prev][''] += 1
+            return stats
+
+    def _generate_mimic_line(self, stats):
+        words = []
+        prev = ''
+        while True:
+            total = 0
+            for _, count in stats[prev].items():
```
Comment (serpis) [inline pladder/log.py]: I think what's going on here is a choice from a list given a weighted distribution. If so, could this function be used? https://docs.python.org/dev/library/random.html#random.choices

Comment (ampleyfly) [reply pladder/log.py]: Yes!

File: pladder/log.py [hunk 2]
```diff
@@ -84,6 +91,12 @@ def SearchLines(self, network, channel, substring, max_count, index):
         else:
             return []
 
+    def Mimic(self, network, channel, nick):
+        if network in self.logdbs:
+            return self.logdbs[network].mimic(channel, nick)
+        else:
+            return []
```
Comment (raek) [inline pladder/log.py]: Doesn't Mimic return a string? If so then shouldn't this line be `return ''`?

File: pladder/log.py [hunk 3]
```diff
@@ -140,6 +153,42 @@ def add_line(self, timestamp, channel, nick, text):
             c.execute("INSERT INTO lines (timestamp, channel, nick, text) VALUES (?, ?, ?, ?);",
                       (timestamp, channel, nick, text))
 
+    def mimic(self, channel, nick):
+        stats = self._generate_mimic_stats(channel, nick)
+        return self._generate_mimic_line(stats)
+
+    def _generate_mimic_stats(self, channel, nick):
```
Comment (raek) [inline pladder/log.py:160]: It will be interesting to see how the performance of this is... :-) Let's find out!

**Review by raek — APPROVED** [review]
Cool!

An idea for optimizing the generation of statistics: Maybe we could keep the statistics dicts in memory in the pladder-log process? The dicts can be polulated at startup with all the history and then they can be updated incrementally from then on.

Anyway, I think it is neat if pladder-log encapsulates all access to the database. Maybe it's the name of pladder-log that should change to reflect what it does?

File: pladder/log.py (hunk 2 above)
Comment (ampleyfly) [reply pladder/log.py]: Good point!

File: pladder/log.py (hunk 3 above)
Comment (ampleyfly) [reply pladder/log.py:160]: I was thinking of caching the statistics somehow, but wanted to try it out first :)

**Pull request merged**